src/authentication.cpp

```diff
@@ -0,0 +1,239 @@
+        auto date_key = irods::s3::authentication::hmac_sha_256(std::string("AWS4").append(secret_key), date);
+        auto date_region_key = irods::s3::authentication::hmac_sha_256(date_key, region);
+        // 'date region service key'
+        // :eyeroll:
+        auto date_region_service_key = irods::s3::authentication::hmac_sha_256(date_region_key, "s3");
+        return irods::s3::authentication::hmac_sha_256(date_region_service_key, "aws4_request");
+    }
+
+    // TODO when this is working, we can improve performance by reusing the same stringstream where possible.
+
+    // Turn the url into the 'canon form'
+    std::string canonicalize_url(const static_buffer_request_parser& request, const boost::urls::url_view& url)
+    {
+        std::stringstream result;
+        // result << request.get().at("Host");
+        for (auto i : url.segments()) {
+            result << '/' << uri_encode(i); // :shrug:
+        }
+        return result.str();
+    }
+
+    bool should_include_header(const std::string_view& sv)
+    {
+        return sv.starts_with("X-Amz") || sv.starts_with("x-amz") || sv == "Host" || sv == "Content-Type" ||
+               sv == "Content-MD5";
+    }
+
+    std::string canonicalize_request(
+        const static_buffer_request_parser& request,
+        const boost::urls::url_view& url,
+        const std::vector<std::string>& signed_headers)
+    {
+        // At various points it wants various fields to be sorted.
+        // so reusing this can at least avoid some of the duplicate allocations and such
+        std::vector<std::string_view> sorted_fields;
+
+        std::stringstream result;
+
+        std::ios state(nullptr);
+        state.copyfmt(result);
+
+        result << request.get().method_string() << '\n';
+        result.copyfmt(result); // re store former formatting
+        result << canonicalize_url(request, url) << '\n';
+        // Canonicalize query string
+        {
+            bool first = true;
+            // Changing it to a pair enables us to sort easily.
+            std::vector<std::pair<std::string, std::string>> params;
+            std::transform(
+                url.encoded_params().begin(), url.encoded_params().end(), std::back_inserter(params), [](auto a) {
+                    if (a.has_value)
+                        return std::make_pair<std::string, std::string>(uri_encode(a.key), uri_encode(a.value));
+                    return std::make_pair(uri_encode(a.key), std::string(""));
+                });
+            std::sort(params.begin(), params.end());
+            for (const auto& param : params) {
+                result << (first ? "" : "&") << param.first;
+                result << '=' << param.second;
+
+                first = false;
+            }
+        }
+        result << '\n';
+
+        for (const auto& header : request.get()) {
+            if (should_include_header(std::string_view(header.name_string().data(), header.name_string().length())))
```

Comment:
What type is returned from `.name_string()`?

Reply:
I'm not sure why I did that, but it compiles without this extra conversion step, so it can go away.